File: packages/runpane/src/doctor.ts
import type { ParsedArgs } from './commands';
import childProcess from 'node:child_process';
import fs from 'node:fs';
import {
  getPaneDaemonEndpoint,
  invokeDaemon,
  resolvePaneDirectory,
  type PaneDaemonEndpoint
} from './daemonClient';
import { resolveExistingPanePath } from './installers';
import { detectPlatform, type PanePlatform } from './platform';
import { resolveRelease } from './releases';
import { getPaneVersion, getWrapperVersion } from './version';

const DOCTOR_DAEMON_TIMEOUT_MS = 5_000;
const DOCTOR_RELEASE_TIMEOUT_MS = 5_000;

interface DaemonDoctorResult {
  ok: true;
  app: {
    version: string;
    isPackaged: boolean;
    platform: string;
    electronVersion?: string;
    nodeVersion?: string;
  };
  daemon: {
    channels: string[];
  };
  repos: {
    count: number;
    active?: {
      id: number;
      name: string;
      path: string;
      active: boolean;
      environment?: string;
      sessionCount: number;
    };
  };
  agentContext: {
    recommendedFirstCommands: string[];
  };
}

interface DoctorReleaseCheck {
  ok: boolean;
  tagName?: string;
  artifactName?: string;
  format?: string;
  preferredDownloadUrl?: string;
  fallbackDownloadUrl?: string;
  error?: string;
}

interface DoctorInstalledPaneCheck {
  found: boolean;
  path?: string;
  version?: string;
}

interface DoctorDaemonCheck {
  reachable: boolean;
  endpoint: PaneDaemonEndpoint;
  result?: DaemonDoctorResult;
  error?: string;
  nextCommand?: string;
}

interface DoctorReport {
  ok: boolean;
  source: 'npm' | 'pip';
  wrapper: {
    runtime: 'node';
    version: string;
    paneDir: string;
    endpoint: PaneDaemonEndpoint;
  };
  platform?: PanePlatform;
  release: DoctorReleaseCheck;
  installedPane: DoctorInstalledPaneCheck;
  daemon: DoctorDaemonCheck;
  remoteSetup: RemoteSetupDoctorCheck;
  nextCommands: string[];
}

interface RemoteSetupDiagnostic {
  code: string;
  severity: 'warning' | 'error';
  message: string;
  recoveryCommand?: string;
}

interface RemoteSetupDoctorCheck {
  ready: boolean;
  displayAvailable: boolean;
  headlessEnvironmentApplied: boolean;
  diagnostics: RemoteSetupDiagnostic[];
}

export async function runDoctor(parsed: ParsedArgs, source: 'npm' | 'pip' = 'npm'): Promise<number> {
  const report = await buildDoctorReport(parsed, source);

  if (parsed.json) {
    console.log(JSON.stringify(report, null, 2));
    return 0;
  }

  renderDoctorText(report);
  return report.release.ok ? 0 : 1;
}

async function buildDoctorReport(parsed: ParsedArgs, source: 'npm' | 'pip'): Promise<DoctorReport> {
  const paneDir = resolvePaneDirectory(parsed.paneDir);
  const endpoint = getPaneDaemonEndpoint(paneDir);
  const platform = collectPlatform();
  const releasePromise = platform.ok
    ? collectReleaseCheck(parsed, source, platform.platform)
    : Promise.resolve({ ok: false, error: platform.error });
  const installedPane = collectInstalledPane(parsed.panePath);
  const daemonPromise = collectDaemonHealth(parsed.paneDir, endpoint);
  const [release, daemon] = await Promise.all([releasePromise, daemonPromise]);
  const remoteSetup = collectRemoteSetupCheck(
    platform.ok ? platform.platform : undefined,
    'format' in release ? release.format : undefined
  );

  return {
    ok: release.ok && daemon.reachable && remoteSetup.ready,
    source,
    wrapper: {
      runtime: 'node',
      version: getWrapperVersion(),
      paneDir,
      endpoint,
    },
    platform: platform.ok ? platform.platform : undefined,
    release,
    installedPane,
    daemon,
    remoteSetup,
    nextCommands: [
      'runpane agent-context --json',
      'runpane agent-context --command "<command>" --json',
      'runpane repos list --json',
    ],
  };
}

interface RemoteSetupProbes {
  displayAvailable: boolean;
  hasFuseRuntime: boolean;
  isRoot: boolean;
  unprivilegedUserNamespaceDisabled: boolean;
  hasSystemctl: boolean;
}

export function collectRemoteSetupCheck(
  platform: PanePlatform | undefined,
  releaseFormat: string | undefined,
  probeOverrides: Partial<RemoteSetupProbes> = {}
): RemoteSetupDoctorCheck {
  if (platform?.os !== 'linux') {
    return {
      ready: true,
      displayAvailable: true,
      headlessEnvironmentApplied: false,
      diagnostics: [],
    };
  }

  const probes: RemoteSetupProbes = {
    displayAvailable: Boolean(process.env.DISPLAY || process.env.WAYLAND_DISPLAY),
    hasFuseRuntime: hasLinuxFuseRuntime(),
    isRoot: typeof process.getuid === 'function' && process.getuid() === 0,
    unprivilegedUserNamespaceDisabled: isUnprivilegedUserNamespaceDisabled(),
    hasSystemctl: commandExists('systemctl'),
    ...probeOverrides,
  };
  const diagnostics: RemoteSetupDiagnostic[] = [];
  if (releaseFormat === 'appimage' && !probes.hasFuseRuntime) {
    diagnostics.push({
      code: 'PANE_APPIMAGE_FUSE_MISSING',
      severity: 'error',
      message: 'The selected AppImage may not start because Pane could not find /dev/fuse and a FUSE mount helper.',
      recoveryCommand: 'Install FUSE for this Linux distribution, or rerun with --format deb on a Debian-based host.',
    });
  }

  if (probes.isRoot) {
    diagnostics.push({
      code: 'PANE_ELECTRON_SANDBOX_ROOT',
      severity: 'error',
      message: 'The Pane Electron runtime should not be launched as root with its sandbox enabled.',
      recoveryCommand: 'Run runpane install daemon as a non-root user.',
    });
  } else if (probes.unprivilegedUserNamespaceDisabled) {
    diagnostics.push({
      code: 'PANE_ELECTRON_SANDBOX_UNAVAILABLE',
      severity: 'error',
      message: 'Unprivileged user namespaces are disabled, so the Electron sandbox may not start.',
      recoveryCommand: 'Enable unprivileged user namespaces for this host, or explicitly use --no-sandbox only if you accept the security tradeoff.',
    });
  }

  if (!probes.hasSystemctl) {
    diagnostics.push({
      code: 'PANE_USER_SERVICE_UNAVAILABLE',
      severity: 'warning',
      message: 'systemctl is unavailable; setup will print a manual daemon command instead of installing a user service.',
    });
  }

  return {
    ready: diagnostics.every((diagnostic) => diagnostic.severity !== 'error'),
    displayAvailable: probes.displayAvailable,
    headlessEnvironmentApplied: true,
    diagnostics,
  };
}

function hasLinuxFuseRuntime(): boolean {
  return fs.existsSync('/dev/fuse') && (commandExists('fusermount') || commandExists('fusermount3'));
}

function isUnprivilegedUserNamespaceDisabled(): boolean {
  try {
    return fs.readFileSync('/proc/sys/kernel/unprivileged_userns_clone', 'utf8').trim() === '0';
  } catch {
    return false;
  }
}

function commandExists(command: string): boolean {
  const result = childProcess.spawnSync('sh', ['-lc', `command -v ${command}`], {
    stdio: 'ignore',
    timeout: 2_000,
  });
  return result.status === 0;
}

function collectPlatform(): { ok: true; platform: PanePlatform } | { ok: false; error: string } {
  try {
    return { ok: true, platform: detectPlatform() };
  } catch (error) {
    return { ok: false, error: errorMessage(error) };
  }
}

async function collectReleaseCheck(
  parsed: ParsedArgs,
  source: 'npm' | 'pip',
  platform: PanePlatform
): Promise<DoctorReleaseCheck> {
  try {
    const release = await resolveRelease({
      version: parsed.paneVersion,
      channel: parsed.channel,
      source,
      platform,
      format: parsed.format,
      target: 'client',
      fetchTimeoutMs: DOCTOR_RELEASE_TIMEOUT_MS,
    });
    return {
      ok: true,
      tagName: release.release.tag_name,
      artifactName: release.artifact.name,
      format: release.format,
      preferredDownloadUrl: release.preferredDownloadUrl,
      fallbackDownloadUrl: release.fallbackDownloadUrl,
    };
  } catch (error) {
    return {
      ok: false,
      error: errorMessage(error),
    };
  }
}

function collectInstalledPane(panePath?: string): DoctorInstalledPaneCheck {
  const installedPath = resolveExistingPanePath(panePath);
  if (!installedPath) {
    return { found: false };
  }

  return {
    found: true,
    path: installedPath,
    version: getPaneVersion(installedPath),
  };
}

async function collectDaemonHealth(paneDir: string | undefined, endpoint: PaneDaemonEndpoint): Promise<DoctorDaemonCheck> {
  try {
    return {
      reachable: true,
      endpoint,
      result: await invokeDaemon<DaemonDoctorResult>('runpane:doctor', [], {
        paneDir,
        timeoutMs: DOCTOR_DAEMON_TIMEOUT_MS,
      }),
    };
  } catch (error) {
    return {
      reachable: false,
      endpoint,
      error: errorMessage(error),
      nextCommand: resolveDaemonRecoveryCommand(endpoint, error),
    };
  }
}

function resolveDaemonRecoveryCommand(endpoint: PaneDaemonEndpoint, error: unknown): string {
  const message = errorMessage(error);
  if (
    endpoint.transport === 'unix'
    && (message.includes('ECONNREFUSED') || fs.existsSync(endpoint.path))
  ) {
    return 'Quit Pane completely, reopen Pane, then rerun runpane doctor --json';
  }

  return 'Open Pane, then rerun runpane doctor --json';
}

function renderDoctorText(report: DoctorReport): void {
  if (report.platform) {
    console.log(`Platform: ${report.platform.os}/${report.platform.arch}`);
  }

  if (report.release.ok) {
    console.log(`Latest release: ${report.release.tagName}`);
    console.log(`Selected artifact: ${report.release.artifactName}`);
    console.log(`Website URL: ${report.release.preferredDownloadUrl}`);
    console.log(`GitHub fallback: ${report.release.fallbackDownloadUrl}`);
  } else {
    console.error(`Release check: failed - ${report.release.error ?? 'unknown error'}`);
  }

  if (report.installedPane.found) {
    console.log(`Installed Pane: ${report.installedPane.path}`);
    console.log(`Installed version: ${report.installedPane.version ?? 'unknown'}`);
  } else {
    console.log('Installed Pane: not found');
  }

  console.log(`Pane directory: ${report.wrapper.paneDir}`);
  console.log(`Daemon endpoint: ${report.daemon.endpoint.transport} ${report.daemon.endpoint.path}`);
  if (report.daemon.reachable) {
    console.log(`Pane daemon: reachable (${report.daemon.result?.repos.count ?? 0} repos)`);
  } else {
    console.log(`Pane daemon: unreachable - ${report.daemon.error ?? 'unknown error'}`);
  }

  console.log(`Remote setup preflight: ${report.remoteSetup.ready ? 'ready' : 'action required'}`);
  if (report.platform?.os === 'linux') {
    console.log(`  Display available: ${report.remoteSetup.displayAvailable ? 'yes' : 'no (headless mode will be applied)'}`);
  }
  for (const diagnostic of report.remoteSetup.diagnostics) {
    console.log(`  ${diagnostic.code}: ${diagnostic.message}`);
    if (diagnostic.recoveryCommand) {
      console.log(`  Recovery: ${diagnostic.recoveryCommand}`);
    }
  }

  console.log('Agent discovery: run "runpane doctor --json" before Pane actions, then "runpane agent-context --json" for full CLI context.');
  console.log('Remote setup: run "runpane setup" for guided setup, or "runpane install daemon --label <name>" for scripting.');
}

function errorMessage(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}

File: packages/runpane-py/src/runpane/doctor.py
from __future__ import annotations

import json
import os
import shutil
from concurrent.futures import ThreadPoolExecutor
from typing import Any, Dict, Optional

from .daemon_client import get_pane_daemon_endpoint, invoke_daemon, resolve_pane_directory
from .installers import resolve_existing_pane_path
from .platforms import PanePlatform, detect_platform
from .releases import resolve_release
from .version import pane_version, wrapper_version

DOCTOR_DAEMON_TIMEOUT_MS = 5_000
DOCTOR_RELEASE_TIMEOUT_SECONDS = 5


def run_doctor(parsed, source: str = "pip") -> int:
    report = build_doctor_report(parsed, source)

    if parsed.json:
        print(json.dumps(without_none(report), indent=2))
        return 0

    render_doctor_text(report)
    return 0 if report["release"]["ok"] else 1


def build_doctor_report(parsed, source: str) -> Dict[str, Any]:
    pane_dir = resolve_pane_directory(parsed.pane_dir)
    endpoint = get_pane_daemon_endpoint(pane_dir)
    platform_result = collect_platform()
    with ThreadPoolExecutor(max_workers=2) as executor:
        release_future = (
            executor.submit(collect_release_check, parsed, source, platform_result["platform"])
            if platform_result["ok"]
            else None
        )
        daemon_future = executor.submit(collect_daemon_health, parsed.pane_dir, endpoint)
        release = (
            release_future.result()
            if release_future
            else {"ok": False, "error": platform_result["error"]}
        )
        daemon = daemon_future.result()
    installed_pane = collect_installed_pane(parsed.pane_path)
    remote_setup = collect_remote_setup_check(
        platform_result.get("platform") if platform_result["ok"] else None,
        release.get("format"),
    )

    return {
        "ok": bool(release["ok"] and daemon["reachable"] and remote_setup["ready"]),
        "source": source,
        "wrapper": {
            "runtime": "python",
            "version": wrapper_version(),
            "paneDir": pane_dir,
            "endpoint": endpoint,
        },
        "platform": platform_to_json(platform_result["platform"]) if platform_result["ok"] else None,
        "release": release,
        "installedPane": installed_pane,
        "daemon": daemon,
        "remoteSetup": remote_setup,
        "nextCommands": [
            "runpane agent-context --json",
            "runpane agent-context --command \"<command>\" --json",
            "runpane repos list --json",
        ],
    }


def collect_remote_setup_check(
    platform: Optional[PanePlatform],
    release_format: Optional[str],
    probe_overrides: Optional[Dict[str, bool]] = None,
) -> Dict[str, Any]:
    if not platform or platform.os != "linux":
        return {
            "ready": True,
            "displayAvailable": True,
            "headlessEnvironmentApplied": False,
            "diagnostics": [],
        }

    probes = {
        "displayAvailable": bool(os.environ.get("DISPLAY") or os.environ.get("WAYLAND_DISPLAY")),
        "hasFuseRuntime": has_linux_fuse_runtime(),
        "isRoot": hasattr(os, "getuid") and os.getuid() == 0,
        "unprivilegedUserNamespaceDisabled": unprivileged_user_namespace_disabled(),
        "hasSystemctl": shutil.which("systemctl") is not None,
    }
    probes.update(probe_overrides or {})

    diagnostics = []
    if release_format == "appimage" and not probes["hasFuseRuntime"]:
        diagnostics.append({
            "code": "PANE_APPIMAGE_FUSE_MISSING",
            "severity": "error",
            "message": "The selected AppImage may not start because Pane could not find /dev/fuse and a FUSE mount helper.",
            "recoveryCommand": "Install FUSE for this Linux distribution, or rerun with --format deb on a Debian-based host.",
        })

    if probes["isRoot"]:
        diagnostics.append({
            "code": "PANE_ELECTRON_SANDBOX_ROOT",
            "severity": "error",
            "message": "The Pane Electron runtime should not be launched as root with its sandbox enabled.",
            "recoveryCommand": "Run runpane install daemon as a non-root user.",
        })
    elif probes["unprivilegedUserNamespaceDisabled"]:
        diagnostics.append({
            "code": "PANE_ELECTRON_SANDBOX_UNAVAILABLE",
            "severity": "error",
            "message": "Unprivileged user namespaces are disabled, so the Electron sandbox may not start.",
            "recoveryCommand": "Enable unprivileged user namespaces for this host, or explicitly use --no-sandbox only if you accept the security tradeoff.",
        })

    if not probes["hasSystemctl"]:
        diagnostics.append({
            "code": "PANE_USER_SERVICE_UNAVAILABLE",
            "severity": "warning",
            "message": "systemctl is unavailable; setup will print a manual daemon command instead of installing a user service.",
        })

    return {
        "ready": all(item["severity"] != "error" for item in diagnostics),
        "displayAvailable": probes["displayAvailable"],
        "headlessEnvironmentApplied": True,
        "diagnostics": diagnostics,
    }


def has_linux_fuse_runtime() -> bool:
    return os.path.exists("/dev/fuse") and bool(shutil.which("fusermount") or shutil.which("fusermount3"))


def unprivileged_user_namespace_disabled() -> bool:
    try:
        with open("/proc/sys/kernel/unprivileged_userns_clone", "r", encoding="utf-8") as handle:
            return handle.read().strip() == "0"
    except OSError:
        return False


def collect_platform() -> Dict[str, Any]:
    try:
        return {"ok": True, "platform": detect_platform()}
    except Exception as error:
        return {"ok": False, "error": str(error)}


def collect_release_check(parsed, source: str, platform: PanePlatform) -> Dict[str, Any]:
    try:
        release = resolve_release(
            version=parsed.pane_version,
            channel=parsed.channel,
            source=source,
            platform=platform,
            format_name=parsed.format,
            target="client",
            fetch_timeout_seconds=DOCTOR_RELEASE_TIMEOUT_SECONDS,
        )
        return {
            "ok": True,
            "tagName": release.release["tag_name"],
            "artifactName": release.artifact["name"],
            "format": release.format,
            "preferredDownloadUrl": release.preferred_download_url,
            "fallbackDownloadUrl": release.fallback_download_url,
        }
    except Exception as error:
        return {
            "ok": False,
            "error": str(error),
        }


def collect_installed_pane(pane_path: Optional[str]) -> Dict[str, Any]:
    installed = resolve_existing_pane_path(pane_path)
    if not installed:
        return {"found": False}

    return {
        "found": True,
        "path": installed,
        "version": pane_version(installed),
    }


def collect_daemon_health(pane_dir: Optional[str], endpoint: Dict[str, str]) -> Dict[str, Any]:
    try:
        return {
            "reachable": True,
            "endpoint": endpoint,
            "result": invoke_daemon("runpane:doctor", [], pane_dir=pane_dir, timeout_ms=DOCTOR_DAEMON_TIMEOUT_MS),
        }
    except Exception as error:
        return {
            "reachable": False,
            "endpoint": endpoint,
            "error": str(error),
            "nextCommand": "Open Pane, then rerun runpane doctor --json",
        }


def render_doctor_text(report: Dict[str, Any]) -> None:
    platform = report.get("platform")
    if platform:
        print(f"Platform: {platform['os']}/{platform['arch']}")

    release = report["release"]
    if release["ok"]:
        print(f"Latest release: {release['tagName']}")
        print(f"Selected artifact: {release['artifactName']}")
        print(f"Website URL: {release['preferredDownloadUrl']}")
        print(f"GitHub fallback: {release['fallbackDownloadUrl']}")
    else:
        print(f"Release check: failed - {release.get('error') or 'unknown error'}")

    installed = report["installedPane"]
    if installed["found"]:
        print(f"Installed Pane: {installed['path']}")
        print(f"Installed version: {installed.get('version') or 'unknown'}")
    else:
        print("Installed Pane: not found")

    daemon = report["daemon"]
    endpoint = daemon["endpoint"]
    print(f"Pane directory: {report['wrapper']['paneDir']}")
    print(f"Daemon endpoint: {endpoint['transport']} {endpoint['path']}")
    if daemon["reachable"]:
        repo_count = ((daemon.get("result") or {}).get("repos") or {}).get("count", 0)
        print(f"Pane daemon: reachable ({repo_count} repos)")
    else:
        print(f"Pane daemon: unreachable - {daemon.get('error') or 'unknown error'}")

    remote_setup = report["remoteSetup"]
    print(f"Remote setup preflight: {'ready' if remote_setup['ready'] else 'action required'}")
    if (report.get("platform") or {}).get("os") == "linux":
        display_status = "yes" if remote_setup["displayAvailable"] else "no (headless mode will be applied)"
        print(f"  Display available: {display_status}")
    for diagnostic in remote_setup["diagnostics"]:
        print(f"  {diagnostic['code']}: {diagnostic['message']}")
        if diagnostic.get("recoveryCommand"):
            print(f"  Recovery: {diagnostic['recoveryCommand']}")

    print('Agent discovery: run "runpane doctor --json" before Pane actions, then "runpane agent-context --json" for full CLI context.')
    print('Remote setup: run "runpane setup" for guided setup, or "runpane install daemon --label <name>" for scripting.')


def platform_to_json(platform: PanePlatform) -> Dict[str, str]:
    return {
        "os": platform.os,
        "arch": platform.arch,
    }


def without_none(value: Any) -> Any:
    if isinstance(value, dict):
        return {key: without_none(item) for key, item in value.items() if item is not None}
    if isinstance(value, list):
        return [without_none(item) for item in value]
    return value
